android/src/main/java/com/reactnativeone/OneModule.java

```diff
@@ -72,62 +83,160 @@ public void init(String siteKey, String touchpointUri, String apiKey,
-            "{\"captures: " + oneResponse.getCaptures().toString() + ", " +
-                  "\"interactionPath\": \"" + oneResponse.getInteractionPath().getValue().toString() + "\", " +
-                  "\"optimizations: " + oneResponse.getOptimizations().toString() + ", " +
-                  "\"statusCode\": " + oneResponse.getHttpStatusCode() + ", " +
-                  "\"tid\": \"" + oneResponse.getTid() + "\", " +
-                  "\"trackers\": " + oneResponse.getTrackers().toString() + "}"
-          );
-        }
 
-        @Override
-        public void onError(@NotNull OneSDKError oneSDKError) {
-          notifyProblem(promise, "" + oneSDKError.getSystemCode(), oneSDKError.getErrorMessage());
+    executor.submit(() -> {
+      OneResponse response;
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+        try {
+          response = One.sendInteraction(THROW_ERRORS, sendInteractionRequest).join();
+          WritableNativeMap responseMap = responseObjectToReadableMap(response);
+          notifyResult(promise, responseMap);
+        } catch (ExecutionException error) {
+          notifyProblem(promise, Integer.toString(error.hashCode()), Objects.requireNonNull(error.getCause()).getLocalizedMessage());
+          Log.e(NAME, "[Thunderhead] Send Interaction Completion Error: " + error.getCause());
+        } catch (OneSDKError error) {
+          notifyProblem(promise, Integer.toString(error.getSystemCode()), error.getLocalizedMessage());
+          Log.e(NAME, "[Thunderhead] Send Interaction SDK Error: " + error.getErrorMessage());
+        } catch (OneAPIError error) {
+          notifyProblem(promise, Integer.toString(error.getHttpStatusCode()), error.getLocalizedMessage());
+          Log.e(NAME, "[Thunderhead] Send Interaction Api Error: " + error.getErrorMessage());
+        } catch (CompletionException error) {
+          notifyProblem(promise, Integer.toString(error.hashCode()), Objects.requireNonNull(error.getCause()).getLocalizedMessage());
+          Log.e(NAME, "[Thunderhead] Send Interaction Completion Error: " + error.getLocalizedMessage());
+        } catch (Exception error) {
+          notifyProblem(promise, Integer.toString(error.hashCode()), error.getLocalizedMessage());
+          Log.e(NAME, "[Thunderhead] Send Interaction Error: " + error.getLocalizedMessage());
         }
-
-        @Override
-        public void onFailure(@NotNull OneAPIError oneAPIError) {
-          notifyProblem(promise, "" + oneAPIError.getHttpStatusCode(), oneAPIError.getErrorMessage());
+      } else {
+        try {
+          response = One.sendInteractionLegacySupport(THROW_ERRORS, sendInteractionRequest).join();
+          WritableNativeMap responseMap = responseObjectToReadableMap(response);
+          notifyResult(promise, responseMap);
+        } catch (ExecutionException error) {
+          notifyProblem(promise, Integer.toString(error.hashCode()), Objects.requireNonNull(error.getCause()).getLocalizedMessage());
```

Comment:
This will produce another exception if the cause is null. Do a normal null check.

```java
Exception cause = error.getCause();
if (cause != null) {
 // notify with cause
} else {
  // notify with something else
}
```

Reply:
do this everywhere

Reply:
I just found out the react promise can take a throwable, so I did it differently.  let me know if you approve.